File: src/main/java/mods/eln/sim/watchdogs/ValueWatchdog.java
package mods.eln.sim.watchdogs;

import mods.eln.*;
import mods.eln.misc.Utils;
import mods.eln.sim.IProcess;

public abstract class ValueWatchdog implements IProcess {

    IDestructible destructible;
    double min;
    double max;

    double timeoutReset = 2;

    double timeout = 0;
    boolean boot = true;
    boolean joker = true;

    double rand = Utils.rand(0.5, 1.5);

    @Override
    public void process(double time) {
        if (boot) {
            boot = false;
            timeout = timeoutReset;
        }
        double value = getValue();
        double overflow = Math.max(value - max, min - value);
        if (overflow > 0) {
            if (joker) {
                joker = false;
                overflow = 0;
            }
        } else {
            joker = true;
        }

        timeout -= time * overflow * rand;
        if (timeout > timeoutReset) {
            timeout = timeoutReset;
        }
        if (timeout < 0) {
            Utils.println("%s destroying %s",
                getClass().getName(),
                destructible.describe());
            if (!Eln.debugExplosions)
                destructible.destructImpl();
        }
    }

    public ValueWatchdog set(IDestructible d) {
        this.destructible = d;
        return this;
    }

    abstract double getValue();

    public void disable() {
        this.max = 100000000;
        this.min = -max;
        this.timeoutReset = 10000000;
    }

    public void reset() {
        boot = true;
    }
}

File: src/main/java/mods/eln/sim/process/destruct/ValueWatchdog.kt
package mods.eln.sim.process.destruct

import mods.eln.Eln
import mods.eln.misc.Utils
import mods.eln.misc.Utils.println
import mods.eln.sim.IProcess

abstract class ValueWatchdog : IProcess {
    private var destructible: IDestructible? = null
    var min = 0.0
    var max = 0.0
    var timeoutReset = 2.0
    var timeout = 0.0
    var boot = true

    // TODO: Rename. Hysteresis?
    private var joker = true
    override fun process(time: Double) {
        if (boot) {
            boot = false
            timeout = timeoutReset
        }
        val value = getValue()
        var overflow = (value - max).coerceAtLeast(min - value)
        if (overflow > 0) {
            if (joker) {
                joker = false
                overflow = 0.0
            }
        } else {
            joker = true
        }
        timeout -= time * overflow * Utils.rand(0.5, 1.5)
        if (timeout > timeoutReset) {
            timeout = timeoutReset
        }
        if (timeout < 0) {
            println(
                "%s destroying %s",
                javaClass.name,
                destructible?.describe()?: "Null destructible"
            )
            if (!Eln.debugExplosions) destructible?.destructImpl()
        }
    }

    fun setDestroys(destructible: IDestructible): ValueWatchdog {
        this.destructible = destructible
        return this
    }

    abstract fun getValue(): Double

    fun reset() {
        boot = true
    }
}
